core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,466 @@
+
+        IndexGraveyardDiff(final StreamInput in) throws IOException {
+            added = Collections.unmodifiableList(in.readList((streamInput) -> new Tombstone(streamInput)));
+            removedCount = in.readVInt();
+        }
+
+        IndexGraveyardDiff(final IndexGraveyard previous, final IndexGraveyard current) {
+            final List<Tombstone> previousTombstones = previous.tombstones;
+            final List<Tombstone> currentTombstones = current.tombstones;
+            final List<Tombstone> added;
+            final int removed;
+            if (previousTombstones.isEmpty()) {
+                // nothing will have been removed, and all entries in current are new
+                added = new ArrayList<>(currentTombstones);
+                removed = 0;
+            } else if (currentTombstones.isEmpty()) {
+                // nothing will have been added, and all entries in previous are removed
+                added = Collections.emptyList();
+                removed = previousTombstones.size();
+            } else {
+                // look through the back, starting from the end, for added tombstones
+                final Tombstone lastAddedTombstone = previousTombstones.get(previousTombstones.size() - 1);
+                final int addedIndex = currentTombstones.lastIndexOf(lastAddedTombstone);
+                if (addedIndex < currentTombstones.size()) {
+                    added = currentTombstones.subList(addedIndex + 1, currentTombstones.size());
+                } else {
+                    added = Collections.emptyList();
+                }
+                // look from the front for the removed tombstones
+                final Tombstone firstTombstone = currentTombstones.get(0);
+                int idx = previousTombstones.indexOf(firstTombstone);
+                if (idx < 0) {
+                    // the first tombstone in the current list wasn't found in the previous list,
+                    // which means all tombstones from the previous list have been deleted.
+                    assert added.equals(currentTombstones); // all previous are removed, so the current list must be the same as the added
+                    idx = previousTombstones.size();
+                }
+                removed = idx;
+            }
+            this.added = Collections.unmodifiableList(added);
+            this.removedCount = removed;
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            out.writeList(added);
+            out.writeVInt(removedCount);
+        }
+
+        @Override
+        public IndexGraveyard apply(final MetaData.Custom previous) {
+            @SuppressWarnings("unchecked")
+            final IndexGraveyard old = (IndexGraveyard) previous;
+            if (removedCount > old.tombstones.size()) {
+                throw new IllegalStateException("IndexGraveyardDiff cannot remove " + removedCount + " entries from " +
```

Comment:
We have a pretty common format of wrapping variables in log messages in brackets. Can we do the same here?